cirq/linalg/decompositions.py

```diff
@@ -249,7 +250,7 @@ def so4_to_magic_su2s(
         if mat.shape != (4, 4) or not predicates.is_special_orthogonal(mat, atol=atol, rtol=rtol):
             raise ValueError('mat must be 4x4 special orthogonal.')
 
-    ab = combinators.dot(MAGIC, mat, MAGIC_CONJ_T)
+    ab = cast(np.ndarray, combinators.dot(MAGIC, mat, MAGIC_CONJ_T))
```

Comment:
Can we avoid the cast if we turn `combinators.dot` into a [function overload](https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading) (so that mypy can infer return type from argument types)?

Reply:
In the current codebase, we could get away with changing cominators.dot to be ndarray -> ndarray, but this would be a change to the function signature (although not the behavior). I've changed it to be ArrayLike -> ndarray which changes the behavior for non-array-inputs slightly by returning the numpy version of each scalar instead of the python version.

This should only cause problems in exceptional circumstances (like if someone is using dot on only scalars and checking the result with isinstance) and like I said there's no evidence in the code-base where dot is called with anything other than arrays.

Reply:
SGTM